### PR TITLE
storage: fast forward kafka updates to high watermark on initialization

### PR DESCRIPTION
**Pull request by petrosagg**
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/source/kafka.rs [hunk 1]
```diff
@@ -1100,8 +1113,15 @@ impl KafkaSourceReader {
...
 
-            let ts = Partitioned::new_singleton(RangeBound::exact(partition), offset);
-            Some((message, ts, 1))
+            let output = message.map(|msg| SourceOutput {
+                key: msg.key,
+                value: msg.value,
+                metadata: msg.metadata,
+                from_time: Partitioned::new_singleton(RangeBound::exact(partition), offset),
+            });
+            let emit_offset = MzOffset::from(std::cmp::max(offset.offset, *partition_as_of));
```
Comment (petrosagg) [inline src/storage/src/source/kafka.rs:1122]: Here we are performing the fast forward action

File: src/storage/src/source/source_reader_pipeline.rs [hunk 2]
```diff
@@ -721,7 +721,7 @@ where
                                     key: message.key,
                                     value: message.value,
                                     metadata: message.metadata,
-                                    from_time: from_ts,
+                                    from_time: message.from_time,
```
Comment (petrosagg) [inline src/storage/src/source/source_reader_pipeline.rs:724]: and here we preserve the original (not fast forwarded timestamp) for everything that comes after reclocking. This is important for upsert to work correctly as it needs to differentiate between updates that happened at different source times

File: src/storage/src/source/kafka.rs (hunk 1 above)
Comment (rjobanp) [reply src/storage/src/source/kafka.rs:1122]:
```suggestion
            // Fast Forwarding the offset we emit at to the max offset in the partition when it was discovered
            let emit_offset = MzOffset::from(std::cmp::max(offset.offset, *partition_as_of));
```

File: src/storage/src/source/kafka.rs [hunk 3]
```diff
@@ -891,7 +895,11 @@ impl KafkaSourceReader {
         self.create_partition_queue(pid, Offset::Offset(start_offset));
 
         let prev = self.last_offsets.insert(pid, start_offset - 1);
+        assert!(prev.is_none());
 
+        let prev = self
+            .partition_as_of
+            .insert(pid, watermarks.high.saturating_sub(1));
```
Comment (rjobanp) [inline src/storage/src/source/kafka.rs:902]: I may be missing something, but how is this re-populated after a restart but once this partition has some offsets committed? Doesn't it reset to the high watermark offset each time the operator starts?

**Review by rjobanp — APPROVED** [review]
Looks good!

**Comment by petrosagg** [comment]
This was an attempt to solve a production issue but we resolved it in another way. Closing

**Pull request closed without merging**